var/spack/repos/builtin/packages/qmcpack/package.py

```diff
@@ -57,31 +59,76 @@ class Qmcpack(CMakePackage):
     # variant('+mixed', default=True, when='+cuda', description="...")
 
     # conflicts
-    conflicts('+soa', when='+cuda')
-    conflicts('^openblas+ilp64')
-    conflicts('^intel-mkl+ilp64')
+    conflicts(
+        '+phdf5',
+        when='~mpi',
+        msg='Parallel collective I/O requires MPI-enabled QMCPACK.\n \
+        Please add \'~phdf5\' to the Spack install line for serial QMCPACK.'
+    )
+    conflicts(
+        '+qe',
+        when='~mpi',
+        msg='QMCPACK QE variant requires MPI due to limitation in QE build system.\n \
+        Please add \'~qe\' to the Spack install line for serial QMCPACK.'
+    )
+    conflicts(
+        '+soa',
+        when='+cuda',
+        msg='QMCPACK SOA variant does not exist for CUDA'
+    )
+    conflicts(
+        '^openblas+ilp64',
+        msg='QMCPACK does not support OpenBLAS 64-bit integer variant'
+    )
+    conflicts(
+        '^intel-mkl+ilp64',
+        msg='QMCPACK does not support MKL 64-bit integer variant'
+    )
+
+    # CUDA conflicts taken from here
+    # https://docs.nvidia.com/cuda/cuda-installation-guide-linux/index.html
+    cuda_warning = 'CUDA is not supported by this version of your compiler'
+    conflicts('%gcc@8:', when='+cuda', msg=cuda_warning)
+    conflicts('%llvm@7:', when='+cuda', msg=cuda_warning)
+    conflicts('%intel@19:', when='+cuda', msg=cuda_warning)
+
+    # QMCPACK 3.6.0 or later requires support for C++14
+    compiler_warning = 'QMCPACK 3.6.0 or later requires a ' \
+                       'compiler with support for C++14'
+    conflicts('%gcc@:4', when='@3.6.0:', msg=compiler_warning)
+    conflicts('%intel@:17', when='@3.6.0:', msg=compiler_warning)
+    conflicts('%pgi@:17', when='@3.6.0:', msg=compiler_warning)
+    conflicts('%llvm@:3.4', when='@3.6.0:', msg=compiler_warning)
 
     # Dependencies match those in the QMCPACK manual.
     # FIXME: once concretizer can unite unconditional and conditional
-    # dependencies the some of the '~mpi' will not be necessary.
-    depends_on('cmake@3.4.3:', type='build')
-    depends_on('mpi', when='+mpi')
+    # dependencies, some of the '~mpi' variants below will not be necessary.
+    # Essential libraries
+    depends_on('cmake@3.4.3:', when='@:3.5.0', type='build')
+    depends_on('cmake@3.6.0:', when='@3.6.0:', type='build')
+    depends_on('boost', when='@:3.5.0')
```

Comment:
I would write this as:
```python
depends_on('boost')
depends_on('boost@1.61.0:', when='@3.6.0:')
```
Spack knows to merge these constraints automatically. There is currently a bug in the concretizer where specifying `qmcpack ^boost@1.61.0` will fail (as things are currently written). This is because Spack hasn't yet decided on the version of `qmcpack`, so it doesn't know whether or not it will depend on `boost`.

Reply:
I can address this.